File: sources/Infrastructure/ObjCExceptions.m
//
//  ObjCExceptions.m
//  iTerm2SharedARC
//
//  Created by George Nachman on 5/25/22.
//

#import "ObjCExceptions.h"

NSError * _Nullable ObjCTryImpl(void (^NS_NOESCAPE block)(void)) {
    @try {
        block();
        return nil;
    } @catch (NSException *e) {
        return [NSError errorWithDomain:e.name
                                   code:0
                               userInfo:@{ NSUnderlyingErrorKey: e,
                                           NSDebugDescriptionErrorKey: e.userInfo ?: @{ },
                                           NSLocalizedFailureReasonErrorKey: (e.reason ?: NSLocalizedStringWithDefaultValue(@"ObjCExceptions.UnknownReason", nil, [NSBundle mainBundle], @"Unknown reason", @"Fallback failure reason when a caught exception has no reason")) }];
    }
}

File: sources/Infrastructure/ObjCExceptions.swift
//
//  ObjCExceptions.swift
//  iTerm2SharedARC
//
//  Created by George Nachman on 5/25/22.
//

import Foundation

func ObjCTry<T>(_ closure: () throws -> T) throws -> T {
    var result: Result<T, Error>? = nil
    let error = ObjCTryImpl {
        do {
            result = .success(try closure())
        } catch {
            result = .failure(error)
        }
    }
    switch result {
    case .success(let value):
        return value
    case .failure(let error):
        throw error
    case .none:
        throw error!
    }
}
